homeassistant/components/rpi_gpio/binary_sensor.py

```diff
@@ -63,12 +74,11 @@ def __init__(self, name, port, pull_mode, bouncetime, invert_logic):
 
         rpi_gpio.setup_input(self._port, self._pull_mode)
 
-        def read_gpio(port):
-            """Read state from GPIO."""
-            self._state = rpi_gpio.read_input(self._port)
-            self.schedule_update_ha_state()
+        def edge_detected(port):
+            """Edge detection handler."""
+            self.hass.add_job(self.async_read_gpio)
 
-        rpi_gpio.edge_detect(self._port, read_gpio, self._bouncetime)
+        rpi_gpio.edge_detect(self._port, edge_detected, self._bouncetime)
```

Comment:
In theory bounce time already processed.
Wouldn't it be better to add two separate callbacks for raising and failing edge?
In that case, you don't need to read the value.
```py
def edge_detect(port, event_callback, bounce):
    """Add detection for RISING and FALLING events."""
    GPIO.add_event_detect(port, GPIO.BOTH, callback=event_callback, bouncetime=bounce)
```

Reply:
Hi sanyatuning,
thanks for reviewing. As you can read out of the long thread in the related 10498 issue this solution is perhaps not the perfect one because still based on RPi.GPIO, gpiozero seems to be better. 
I agree that reading the value **could** be avoided, but this add more reliability at low cost. Some of the users in the issue thread are reporting missed edged and in this case reading out the value seems to be the only workaround.
At least this solution is tested for month by many users (including me) in a custom component and it works way better then the current implementation on master.
So I would like to keep it like it is.